presto/logging.py

```diff
@@ -14,12 +14,65 @@
 from __future__ import print_function
 
 import logging
+import inspect
+import string
 
 LEVEL = logging.INFO
 
+class LoggingFormatError(Exception):
+    """
+    Raise this exception when the logging adapter is unable to
+    deduce whether arguments are meant for formatting or for the logger itself.
+    """
+
+class LogMessage:
+    def __init__(self, fmt, args, kwargs):
+        # fmt is assumed to be str, performed in FormatLogger.log()
+        self.fmt = fmt
+        self.args = args
+        self.kwargs = kwargs
+
+    def __str__(self):
+        return self.fmt.format(*self.args, **self.kwargs)
+
+    @property
+    def keywords(self):
+        return 
+
+class FormatLogger(logging.LoggerAdapter):
+    def __init__(self, logger, extra=None):
+        super(FormatLogger, self).__init__(logger, extra or {})
+
+        # The keywords may be different per instance of FormatLogger depending 
+        # on whether this logger is derived and _log is overridden.
+        sig = inspect.signature(self.logger._log)
+        self.logger_keywords = set(p.name for p in sig.parameters.values() if p.default != inspect.Parameter.empty)
+    
+    def filter_kwargs(self, msg, kwargs):
+        msg_keywords = set(fname for _, fname, _, _ in string.Formatter().parse(msg) if fname)
+
+        ambiguous_parameters = self.logger_keywords & msg_keywords
+        if ambiguous_parameters:
+            raise LoggingFormatError("Ambiguous parameters {} used during logging.".format(ambiguous_parameters))
+
+        logger_kwargs = {}
+        msg_kwargs = {}
+        for name, val in kwargs.items():
+            if name in logger_kwargs:
+                logger_kwargs[name] = val
+            else:
+                msg_kwargs[name] = val
+        return logger_kwargs, msg_kwargs
+
+    def log(self, level, msg, *args, **kwargs):
+        if self.isEnabledFor(level):
+            msg = str(msg)
+            msg, kwargs = self.process(msg, kwargs)
+            logger_kwargs, msg_kwargs = self.filter_kwargs(msg, kwargs)
+            self.logger._log(level, LogMessage(msg, args, msg_kwargs), tuple(), **logger_kwargs)
 
 # TODO: provide interface to use ``logging.dictConfig``
 def get_logger(name, log_level=LEVEL):
     logger = logging.getLogger(name)
     logger.setLevel(log_level)
-    return logger
+    return FormatLogger(logger)
```

Comment:
What are the builtin `logging`'s formatting capabilities? 
Is there something we could use to avoid adding a wrapper?

Reply:
The builtin logging's simplest formatting capabilities can allow you to lazily log in the old python2 syntax:

```
logging.log("%s", something)
```

There currently is no support for python 3 formatting due to ambiguous kwargs that might arise. In general, python2 style is [not encouraged](https://docs.python.org/3/library/stdtypes.html?highlight=sprintf#printf-style-string-formatting). @ wrapper, the `LoggingAdapter` is actually a pattern that python suggests on using - we're not actually adding any overhead because all I've done is replace the string processing function on the logger level :)

Reply:
IMo the old-school `%s` formatting is good enough for our modest needs.
I would rather keep the code simple using what is immediately available from the standard library -- and wait for the standard library to gain more reach formatting options.

Reply:
sure, I can open another PR for that instead 👍